File: zivid_samples/src/sample_capture_and_save.cpp
// Copyright 2024 Zivid AS
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the Zivid AS nor the names of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include <exception>
#include <filesystem>

#include <rclcpp/rclcpp.hpp>

#include <zivid_interfaces/srv/capture_and_save.hpp>

/*
 * This sample shows how to set the settings_yaml parameter of the zivid node, then invoke the
 * capture_and_save service, and read the response from the service call. If successful, the
 * captured frame is stored in a temporary directory.
 */

void set_settings(const std::shared_ptr<rclcpp::Node> & node)
{
  RCLCPP_INFO(node->get_logger(), "Setting parameter 'settings_yaml'");
  const std::string settings_yml =
    R"(
__version__:
  serializer: 1
  data: 22
Settings:
  Acquisitions:
    - Acquisition:
        Aperture: 5.66
        ExposureTime: 8333
  Processing:
    Filters:
      Outlier:
        Removal:
          Enabled: yes
          Threshold: 5
)";

  auto param_client = std::make_shared<rclcpp::AsyncParametersClient>(node, "zivid_camera");
  while (!param_client->wait_for_service(std::chrono::seconds(3))) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(node->get_logger(), "Client interrupted while waiting for service to appear.");
      std::terminate();
    }
    RCLCPP_INFO(node->get_logger(), "Waiting for the parameters client to appear...");
  }

  auto result = param_client->set_parameters({rclcpp::Parameter("settings_yaml", settings_yml)});
  if (rclcpp::spin_until_future_complete(node, result) != rclcpp::FutureReturnCode::SUCCESS) {
    RCLCPP_ERROR(node->get_logger(), "Failed to set `settings_yaml` parameter");
    std::terminate();
  }
}

void capture_and_save(const std::shared_ptr<rclcpp::Node> & node)
{
  using zivid_interfaces::srv::CaptureAndSave;

  auto client = node->create_client<CaptureAndSave>("capture_and_save");
  while (!client->wait_for_service(std::chrono::seconds(3))) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(node->get_logger(), "Client interrupted while waiting for service to appear.");
      std::terminate();
    }
    RCLCPP_INFO(node->get_logger(), "Waiting for the capture_and_save service to appear...");
  }

  const std::string filename = "zivid_sample_capture_and_save.zdf";

  auto request = std::make_shared<CaptureAndSave::Request>();
  request->file_path = (std::filesystem::temp_directory_path() / filename).string();
  RCLCPP_INFO(
    node->get_logger(), "Sending capture and save request with file path: %s",
    request->file_path.c_str());

  auto result = client->async_send_request(request);
  if (rclcpp::spin_until_future_complete(node, result) != rclcpp::FutureReturnCode::SUCCESS) {
    RCLCPP_ERROR(node->get_logger(), "Failed to call capture_and_save service");
    std::terminate();
  }

  auto capture_response = result.get();
  if (!capture_response->success) {
    RCLCPP_ERROR(
      node->get_logger(), "Capture and save operation was unsuccessful: %s",
      capture_response->message.c_str());
    std::terminate();
  }

  RCLCPP_INFO(node->get_logger(), "Capture and save operation successful");
}

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("sample_capture_and_save");
  RCLCPP_INFO(node->get_logger(), "Started the sample_capture_and_save node");

  set_settings(node);

  capture_and_save(node);

  RCLCPP_INFO(node->get_logger(), "Spinning node.. Press Ctrl+C to abort.");
  rclcpp::spin(node);

  rclcpp::shutdown();

  return EXIT_SUCCESS;
}

File: zivid_samples/scripts/sample_capture_and_save.py
#!/usr/bin/env python

import sys
import tempfile

from rcl_interfaces.srv import SetParameters
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from rclpy.parameter import Parameter
from zivid_interfaces.srv import CaptureAndSave


class Sample(Node):

    def __init__(self):
        super().__init__('sample_capture_and_save_py')

        self.capture_and_save_service = self.create_client(
            CaptureAndSave, 'capture_and_save'
        )
        while not self.capture_and_save_service.wait_for_service(timeout_sec=3.0):
            self.get_logger().info(
                'capture_and_save service not available, waiting again...'
            )

        self._set_settings()

    def _set_settings(self):
        self.get_logger().info('Setting parameter `settings_yaml`')
        settings_parameter = Parameter(
            'settings_yaml',
            Parameter.Type.STRING,
            """
__version__:
  serializer: 1
  data: 22
Settings:
  Acquisitions:
    - Acquisition:
        Aperture: 5.66
        ExposureTime: 8333
  Processing:
    Filters:
      Outlier:
        Removal:
          Enabled: yes
          Threshold: 5
""",
        ).to_parameter_msg()

        param_client = self.create_client(SetParameters, 'zivid_camera/set_parameters')
        while not param_client.wait_for_service(timeout_sec=3):
            self.get_logger().info('Parameter service not available, waiting again...')

        future = param_client.call_async(
            SetParameters.Request(parameters=[settings_parameter])
        )
        rclpy.spin_until_future_complete(self, future, timeout_sec=30)
        if not future.result():
            raise RuntimeError('Failed to set parameters')

    def capture(self):
        file_path = tempfile.gettempdir() + '/zivid_sample_capture_and_save.zdf'
        self.get_logger().info(
            f'Calling capture_and_save service with file path: {file_path}'
        )
        request = CaptureAndSave.Request(file_path=file_path)
        return self.capture_and_save_service.call_async(request)


def main(args=None):
    rclpy.init(args=args)

    try:
        sample = Sample()
        future = sample.capture()

        rclpy.spin_until_future_complete(sample, future)

        response: CaptureAndSave.Response = future.result()
        if not response.success:
            sample.get_logger().error(f'Failed capture and save: {response.message}')

        sample.get_logger().info('Capture and save complete')

        sample.get_logger().info('Spinning node.. Press Ctrl+C to abort.')
        rclpy.spin(sample)

    except KeyboardInterrupt:
        pass
    except ExternalShutdownException:
        sys.exit(1)


if __name__ == '__main__':
    main()
